File: examples/ch12/py/concept_composite.py
"""Composite — canonical conceptual example (runnable).

A Component interface; a Leaf implements it directly; a Composite holds children
(each itself a Component) and delegates the operation to them, aggregating the
results. The client treats leaf and composite identically.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class Component(ABC):
    @abstractmethod
    def operation(self) -> int: ...


class Leaf(Component):
    def __init__(self, value: int) -> None:
        self._value = value

    def operation(self) -> int:
        return self._value


class Composite(Component):
    def __init__(self) -> None:
        self._children: list[Component] = []

    def add(self, child: Component) -> None:
        self._children.append(child)

    def operation(self) -> int:
        return sum(child.operation() for child in self._children)  # aggregate


if __name__ == "__main__":
    root = Composite()
    root.add(Leaf(1))
    branch = Composite()
    branch.add(Leaf(2))
    branch.add(Leaf(3))
    root.add(branch)            # one real nested level
    print(root.operation())     # 6

File: examples/ch12/ts/conceptComposite.ts
// Composite — canonical conceptual example. Run: npx tsx conceptComposite.ts
//
// A Component interface; a Leaf implements it directly; a Composite holds
// children (each itself a Component) and aggregates their results. The client
// treats leaf and composite identically.

export interface Component {
  operation(): number;
}

export class Leaf implements Component {
  constructor(private readonly value: number) {}
  operation(): number {
    return this.value;
  }
}

export class Composite implements Component {
  private readonly children: Component[] = [];
  add(child: Component): void {
    this.children.push(child);
  }
  operation(): number {
    return this.children.reduce((sum, c) => sum + c.operation(), 0); // aggregate
  }
}

// Demo
const root = new Composite();
root.add(new Leaf(1));
const branch = new Composite();
branch.add(new Leaf(2));
branch.add(new Leaf(3));
root.add(branch); // one real nested level
console.log(root.operation()); // 6
